src/timing.c

```diff
@@ -113,18 +135,124 @@ void jl_destroy_timing(void)
     jl_ptls_t ptls = jl_current_task->ptls;
     jl_timing_block_t *stack = ptls->timing_stack;
     while (stack) {
-        _jl_timing_block_destroy(stack);
+        jl_timing_block_end(stack);
         stack = stack->prev;
     }
 }
 
-jl_timing_block_t *jl_pop_timing_block(jl_timing_block_t *cur_block)
+static const int get_timing_subsystem(const char *subsystem) {
+    const char **match = (const char **)bsearch(
+        subsystem, jl_timing_subsystems, JL_TIMING_SUBSYSTEM_LAST,
+        sizeof(const char *), (int(*)(const void*, const void*))strcmp
+    );
+    if (!match)
+        return JL_TIMING_SUBSYSTEM_LAST;
+
+    return (int)(match - &jl_timing_subsystems[0]);
+}
+
+// This function is analogous to __itt_event_create but for the counts backend
+//
+// `event` is required to live forever
+JL_DLLEXPORT jl_timing_counts_event_t *jl_timing_counts_event_create(const char *event) {
+#ifdef USE_TIMING_COUNTS
+    const size_t n = jl_timing_counts_events.len;
+    for (size_t i = 0; i < n; i++) {
+        jl_timing_counts_event_t *other_event = (jl_timing_counts_event_t *)jl_timing_counts_events.items[i];
+        if (strcmp(event, other_event->name) == 0)
+            return other_event;
+    }
```

Comment:
I would like to see this be a hash table lookup or something faster than linear time.

Reply:
I don't think this event-creation overhead is going to matter much until we start creating many more events (we have less than 100 now)

Can that be tackled in a follow-up?

Reply:
Yeah we can push this to a follow up.